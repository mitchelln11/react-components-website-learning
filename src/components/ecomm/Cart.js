

export function CartItem() {
    return(
        <div id="cart-page">
            <div className="cart-row-container">
                <div className="cart-item">
                    <div className="item-details">
                        {/* <img src={productList[id].image || placeholderImg} alt={productList[id].productName} /> */}
                        <img src={require('../../images/merrell-moab-mens-hiking-boot.jpg')} alt="Merrell Moab Men's Hiking Boot" />
                        <h2>Merrell Moab Men's Hiking Boot</h2>
                    </div>
                    <div className="price-details">
                        <p>$65.00</p>
                        <p>Sale: $55.00</p>
                    </div>
                </div>

                <div className="cart-item">
                    <div className="item-details">
                        {/* <img src={productList[id].image || placeholderImg} alt={productList[id].productName} /> */}
                        <img src={require('../../images/merrell-moab-mens-hiking-boot.jpg')} alt="Merrell Moab Men's Hiking Boot" />
                        <h2>Merrell Moab Men's Hiking Boot</h2>
                    </div>
                    <div className="price-details">
                        <p>$65.00</p>
                        <p>Sale: $55.00</p>
                    </div>
                </div>

                <div className="cart-item">
                    <div className="item-details">
                        {/* <img src={productList[id].image || placeholderImg} alt={productList[id].productName} /> */}
                        <img src={require('../../images/merrell-moab-mens-hiking-boot.jpg')} alt="Merrell Moab Men's Hiking Boot" />
                        <h2>Merrell Moab Men's Hiking Boot</h2>
                    </div>
                    <div className="price-details">
                        <p>$65.00</p>
                        <p>Sale: $55.00</p>
                    </div>
                </div>

                <div className="cart-item">
                    <div className="item-details">
                        {/* <img src={productList[id].image || placeholderImg} alt={productList[id].productName} /> */}
                        <img src={require('../../images/merrell-moab-mens-hiking-boot.jpg')} alt="Merrell Moab Men's Hiking Boot" />
                        <h2>Merrell Moab Men's Hiking Boot</h2>
                    </div>
                    <div className="price-details">
                        <p>$65.00</p>
                        <p>Sale: $55.00</p>
                    </div>
                </div>

                <div className="cart-item">
                    <div className="item-details">
                        {/* <img src={productList[id].image || placeholderImg} alt={productList[id].productName} /> */}
                        <img src={require('../../images/merrell-moab-mens-hiking-boot.jpg')} alt="Merrell Moab Men's Hiking Boot" />
                        <h2>Merrell Moab Men's Hiking Boot</h2>
                    </div>
                    <div className="price-details">
                        <p>$65.00</p>
                        <p>Sale: $55.00</p>
                    </div>
                </div>

                <div className="cart-item">
                    <div className="item-details">
                        {/* <img src={productList[id].image || placeholderImg} alt={productList[id].productName} /> */}
                        <img src={require('../../images/merrell-moab-mens-hiking-boot.jpg')} alt="Merrell Moab Men's Hiking Boot" />
                        <h2>Merrell Moab Men's Hiking Boot</h2>
                    </div>
                    <div className="price-details">
                        <p>$65.00</p>
                        <p>Sale: $55.00</p>
                    </div>
                </div>

                <div className="cart-item">
                    <div className="item-details">
                        {/* <img src={productList[id].image || placeholderImg} alt={productList[id].productName} /> */}
                        <img src={require('../../images/merrell-moab-mens-hiking-boot.jpg')} alt="Merrell Moab Men's Hiking Boot" />
                        <h2>Merrell Moab Men's Hiking Boot</h2>
                    </div>
                    <div className="price-details">
                        <p>$65.00</p>
                        <p>Sale: $55.00</p>
                    </div>
                </div>

                <div className="cart-item">
                    <div className="item-details">
                        {/* <img src={productList[id].image || placeholderImg} alt={productList[id].productName} /> */}
                        <img src={require('../../images/merrell-moab-mens-hiking-boot.jpg')} alt="Merrell Moab Men's Hiking Boot" />
                        <h2>Merrell Moab Men's Hiking Boot</h2>
                    </div>
                    <div className="price-details">
                        <p>$65.00</p>
                        <p>Sale: $55.00</p>
                    </div>
                </div>

                <div className="cart-item">
                    <div className="item-details">
                        {/* <img src={productList[id].image || placeholderImg} alt={productList[id].productName} /> */}
                        <img src={require('../../images/merrell-moab-mens-hiking-boot.jpg')} alt="Merrell Moab Men's Hiking Boot" />
                        <h2>Merrell Moab Men's Hiking Boot</h2>
                    </div>
                    <div className="price-details">
                        <p>$65.00</p>
                        <p>Sale: $55.00</p>
                    </div>
                </div>

            {/* Potential items to add to page */}
            {/* Quantity checkbox */}
            {/* Delete option */}
            {/* Go to item on click */}
            </div>

            {/* Shipping info Sidebar */}
            <aside id="cart-sidebar" className="primary-color primary-font">
                <div className="price-totals">
                    <p>Item Total:</p>
                    <p>Tax</p>
                </div>
            </aside>
        </div>
    )
}