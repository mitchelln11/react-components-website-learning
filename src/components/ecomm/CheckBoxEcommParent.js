import { useState  } from "react";
import { productList } from '../../dataSamples/eCommProducts';
import placeholderImg from '../../images/product-placeholder.jpg';

// return individual values without duplicates
export const eCommValues = {
    id: [...new Set(productList.map(product => product.id))],
    name: [...new Set(productList.map(product => product.productName))],
    price: [...new Set(productList.map(product => product.price))],
    brand: [...new Set(productList.map(product => product.brand))],
    mainCategory: [...new Set(productList.map(product => product.mainCategory))],
    labels: [...new Set(productList.map(product => product.labels))],
    color: [...new Set(productList.map(product => product.color))],
    image: [...new Set(productList.map(product => product.image))]
}

// Try referencing this article on sharing component states
// https://beta.reactjs.org/learn/sharing-state-between-components
// !!!!!!Need to Lift State Up!!!!!!
// function findCheckbox(desiredItem) {
//     let checkboxFound = document.getElementById(desiredItem)
//     console.log(checkboxFound)
// }

// Outer shell to the Checkbox fields
function GenerateFieldSet({ChecklistTitle, children}){
    return (
    <fieldset>
        <legend>{ChecklistTitle}</legend>
            {children}
    </fieldset>
    )
}

// Generate individual checkbox item input
function GenerateCheckListItem({brand}) {
    const [isChecked, setIsChecked] = useState(false);

    function handleOnChange(e) {
        setIsChecked(e.target.checked)
    }

    // findCheckbox("Merrell")

    return (
        <div className="checkbox-item">
            <input 
                id={brand}
                type="checkbox"
                name={brand}
                value={brand}
                checked={isChecked}
                onChange={handleOnChange}
            />
            <label htmlFor={brand}>{brand}</label>
        </div>
    )
}



export function StoreSection() {
    // Display all options without duplicates
    function DisplayAllChecklistItems({prodVal}) {
        return (
            eCommValues[prodVal].map((item, index) => {
                return <GenerateCheckListItem key={index} brand={item} />
        })
    )}

    // function ProductGallery() {
    //     console.log("Is this registering?")
    //     return(
    //         productList.map((product) => {
    //             return (
    //                 <h3>{product.brand}</h3>
    //                 // <DisplaySingleProduct key={product.id} productName={product.productName} price={product.price} />
    //             )
    //         })
    //     )
    // }


    // All the UI items
    return (
        <div id="product-store">
            <aside>
            <h3>Filters</h3>
            <GenerateFieldSet ChecklistTitle="Brands">
                <DisplayAllChecklistItems prodVal="brand" />
            </GenerateFieldSet>
            <GenerateFieldSet ChecklistTitle="Categories">
                <DisplayAllChecklistItems prodVal="mainCategory" />
            </GenerateFieldSet>
            </aside>
            <div id="product-gallery">
                <ProductGallery />
            </div>
        </div>
    )
}

/////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////
//////////----------STORE SECTION-----------------///////////////
/////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////

function DisplaySingleProduct({id, productName, price, image}) {
    return (
        <div id={`product${id}`} className="product-tab" key={`product${id}`}>
            <h3>{productName}</h3>
            <h4>${price.toFixed(2)}</h4>
            <img src={image} alt={productName + id} />
        </div>
    )
}

// The actual UI logic
export function ProductGallery() {

    // This works, just commenting for now
    // function findProduct() {
    //     console.log(productList.filter(item => item.brand === prodVal));
    // }

    // findProduct({prodVal})

    // This should be the else statement for when nothing is checked
    return (
        productList.map((product) => {
            return (
                <DisplaySingleProduct key={product.id} productName={product.productName} price={product.price} image={product.image || placeholderImg} id={product.id} />
            )
        })
    )
}

