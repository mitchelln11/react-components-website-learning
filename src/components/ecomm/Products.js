import placeholderImg from '../../images/product-placeholder.jpg';
import { productList } from '../../dataSamples/eCommProducts';

// return individual values without duplicates
export const eCommValues = {
    id: [...new Set(productList.map(product => product.id))],
    name: [...new Set(productList.map(product => product.productName))],
    price: [...new Set(productList.map(product => product.price))],
    brand: [...new Set(productList.map(product => product.brand))],
    mainCategory: [...new Set(productList.map(product => product.mainCategory))],
    labels: [...new Set(productList.map(product => product.labels))],
    color: [...new Set(productList.map(product => product.color))],
    image: [...new Set(productList.map(product => product.image))]
}

function DisplaySingleProduct({id, productName, price, image}) {
    return (
        <div id={`product${id}`} className="product-tab" key={`product${id}`}>
            <h3>{productName}</h3>
            <h4>${price.toFixed(2)}</h4>
            <img src={image} alt={productName + id} />
        </div>
    )
}

// The actual UI logic
export function ProductGallery() {
    return (
        productList.map((product) => {
            return (
                <DisplaySingleProduct key={product.id} productName={product.productName} price={product.price} image={product.image || placeholderImg} id={product.id} />
            )
        })
    )
}