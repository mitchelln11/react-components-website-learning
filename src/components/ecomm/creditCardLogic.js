import { useState, useRef } from "react";
import { formatCreditCard, getCreditCardType } from "cleave-zen";

// export function parseCcForType(ccNum) {
//     switch (true) {
//         case ccNum.charAt(0) === "2":
//             return "Mastercard"
//         case ccNum.charAt(0) === "3":
//             return "American Express"
//         case ccNum.charAt(0) === "4":
//             return "Visa"
//         case ccNum.charAt(0) === "5":
//             return "Mastercard"
//         case (ccNum.charAt(0) === "6"):
//             return "Discover"
//         default: 
//             return "No Credit Card Match"
//     }
// }

export default function CcLogic({fieldId}) {
    const inputRef = useRef(null)
    const [ ccValue, setCcValue ] = useState('');
    const [ ccType, setCctype ] = useState('');

    const handleCcChange = (event) => {
        const value = event.target.value;
        setCcValue(formatCreditCard(value));
        setCctype(getCreditCardType(value));
    }
    
    return (
        <>
            <input 
                ref={inputRef}
                value={ccValue}
                id={fieldId} 
                name={fieldId} 
                onChange={handleCcChange} 
                maxlength="19" 
                type="text" 
                inputmode="numeric" 
                pattern="[0-9]{4} [0-9]{4} [0-9]{4} [0-9]{4}" 
                placeholder="Ex: 4502783790218767" 
            />
            <div>type: {ccType} </div>
        </>
    )   
}