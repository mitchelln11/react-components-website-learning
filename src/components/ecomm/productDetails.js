import placeholderImg from '../../images/product-placeholder.jpg';
import { useParams } from 'react-router-dom';
import { productList } from '../../dataSamples/eCommProducts';

export default function Product() {
    // parameter used in 'useParams' must match the way it's spelled on the Route
    const { id } = useParams();

    return (
        <div id="product-page">
            <div id="products">
                <div className="image-section">
                    <img src={productList[id].image || placeholderImg} alt="Placeholder" />
                </div>
                <div className="product-info">
                    <h1>Product Page</h1>
                    <div className="additional-info"> 
                        <div>
                            <h2>{productList[id].price}</h2> 
                        </div>
                        <div>
                            <h3>{productList[id].brand}</h3>
                            <h4>{productList[id].color}</h4>
                            {productList[id].description}
                        </div>
                    </div>
                </div>
            </div>

            <footer>
                <h2>Tags</h2>
                <ul className="product-tags">
                    <li>footwear</li>
                    <li>men's</li>
                    <li>boot</li>
                    <li>hiking</li>
                    <li>camping</li>
                </ul>
                <p>These are hard-coded, but I could loop through the product labels in the future.</p>
            </footer>
        </div>
    )
}