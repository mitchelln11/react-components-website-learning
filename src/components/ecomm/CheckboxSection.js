import { eCommValues, GenerateCheckboxList } from './CheckBoxEcommParent';

// Display all options without duplicates
export default function GenericCheckboxList({prodVal}) {
    return (
        eCommValues[prodVal].map((item, index) => {
            return <GenerateCheckboxList key={index} checkboxValue={item} data-product-element={prodVal} />
    })
)}