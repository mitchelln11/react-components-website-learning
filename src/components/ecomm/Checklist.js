import { useState  } from "react";
import { eCommValues } from './Products'

// Generate individual checkbox item input
function GenerateCheckListItem({brand}) {
    const [isChecked, setIsChecked] = useState(false);

    function handleOnChange(e) {
        setIsChecked(e.target.checked);
        // retrieveCheckedProducts();
    }

    return (
        <div className="checkbox-item">
            <input 
                id={brand}
                type="checkbox"
                name={brand}
                value={brand}
                checked={isChecked}
                onChange={handleOnChange}
            />
            <label htmlFor={brand}>{brand}</label>
        </div>
    )
}

// Display all options without duplicates
export function DisplayAllChecklistItems({prodVal}) {
    return (
        eCommValues[prodVal].map((item, index) => {
            return <GenerateCheckListItem key={index} brand={item} />
    })
)}