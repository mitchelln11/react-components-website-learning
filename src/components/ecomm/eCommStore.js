import { useState  } from "react";
import { DisplayAllChecklistItems } from './Checklist';
import { ProductGallery } from './Products';
// import { productList } from '../../dataSamples/eCommProducts';

// Outer shell to the Checkbox fields
function GenerateFieldSet({ChecklistTitle, children}){
    return (
    <fieldset>
        <legend>{ChecklistTitle}</legend>
            {children}
    </fieldset>
    )
}

// What needs to happen?
// 1. Start with empty array and full productList
// 2. When clicked, get the value of the checkbox
// 3. Add value of checkbox to empty array
// 4. Return new array
// 5. Run new array against product to display only checked items
// 6. If unchecked, get value of checkbox
// 7. Run checkbox value against array
//  7b. If it currently exists in array, remove it
// 8. Run newly, shortened array against products to show new, selected items
// 9. If none selected, show all
// function reducer(state, action) {
//     switch (action.type) {
//         case 'SET_CHECKED':
//             return {
//                 ...state, 
//                 checked: action.payload,
//             };
//         case 'VALUE_TO_ARRAY':
//             return {
//                 ...state, 
//                 selectedCeckboxes: action.payload,
//             };
//         default:
//             throw new Error("What the fudge u trying to do?");
//     }
// }

export function StoreSection() {
    const [isChecked, setIsChecked] = useState(false);
    const [checkedCount, setCheckedCount] = useState([]);

    // const [state, dispatch] = useReducer(reducer, {
    //     checked: false,
    //     selectedCeckboxes: []
    // });

    function handleOnChange(e) {
        setIsChecked(e.target.checked);

        // dispatch({ type: 'checked_box', payload: e.target.value})

        e.target.checked ?
        setCheckedCount([...checkedCount, e.target.value])
        :
        setCheckedCount(checkedCount.filter((count) => count !== e.target.value))
    }

    // All the UI items
    return (
        <div id="product-store">
            <aside>
            <h3>Filters</h3>
            <GenerateFieldSet ChecklistTitle="Brands">
                <DisplayAllChecklistItems prodVal="brand" isChecked={isChecked} onChange={handleOnChange} />
            </GenerateFieldSet>
            <GenerateFieldSet ChecklistTitle="Categories">
                <DisplayAllChecklistItems prodVal="mainCategory" isChecked={isChecked} onChange={handleOnChange} />
            </GenerateFieldSet>
            </aside>
            <div id="product-gallery">
                <ProductGallery checkedArray={checkedCount} /*dispatch={dispatch}*/ />
            </div>
        </div>
    )
}