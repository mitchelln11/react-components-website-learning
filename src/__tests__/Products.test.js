import { render, screen } from '@testing-library/react'
import { DisplaySingleProduct, ProductGallery } from '../components/ecomm/Products';
import { BrowserRouter } from 'react-router-dom';
import placeholderImg from '../images/product-placeholder.jpg';

const productEx = {
    id: 1,
    productName: "Roof Rack",
    price: 115.00,
    brand: "Saris",
    mainCategory: "Storage",
    description: "The number one storage rack for road-trip storage.",
    labels: [
        "storage",
        "vehicle"
    ],
    color: "black",
    image: placeholderImg
}

describe('DisplaySingleProduct', () => {
    const MockLink = () => {
        return (
            <BrowserRouter>
                <DisplaySingleProduct {...productEx} />
            </BrowserRouter>
        )
    }

    test('Product headers display', () => {
        render(<MockLink />);
        const product = screen.getAllByRole('heading');
        expect(product).toMatchSnapshot()
    })

    test('Product image displays', () => {
        render(<MockLink />);
        const img = screen.getByAltText(/Roof Rack/)
        expect(img).toBeInTheDocument();
    })
})

describe('ProductGallery', () => {
    const MockLink = (mockedChecked) => {
        return (
            <BrowserRouter>
                <ProductGallery {...mockedChecked} />
            </BrowserRouter>
        )
    }

    test('Product with ID of 2 displays', () => {
        render(<MockLink checkedCount={2} />);
        const product = screen.getByAltText(/Natural Walking Stick/)
        expect(product).toBeInTheDocument();
    })

    test('Multiple products display if nothing is selected', () => {
        render(<MockLink checkedCount={0} />);
        const products = screen.getAllByRole('heading', { level: 3} )
        expect(products).toHaveLength(4)
    })
})