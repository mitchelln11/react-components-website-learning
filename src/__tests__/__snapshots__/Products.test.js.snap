// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplaySingleProduct Product headers display 1`] = `
Array [
  <h3>
    Roof Rack
  </h3>,
  <h4>
    $
    115.00
  </h4>,
]
`;
